FeatureEngine.py

```diff
@@ -42,37 +44,243 @@ def map_labels_to_invoice_OCR(self, data_path: str, verbose: bool = False):
-            token, date_related_tokens
+        # ave dist to neighbours (pixel and relative)
+        min_dist_neighbours = [
+            calc_min_dist(t, token) for t in invoicePage.grouped_tokens
+        ]
+        features["average_dist_neighbours_pixel"] = sum(min_dist_neighbours) / len(
+            min_dist_neighbours
+        )
+        invoice_diag = math.sqrt(
+            invoicePage.size["y"] ** 2 + invoicePage.size["x"] ** 2
+        )
+        features["average_dist_neighbours_rel"] = (
+            features["average_dist_neighbours_pixel"] / invoice_diag
+        )
+        N = 5  # N is arbitrary
+        min_dist_neighbours.sort()
+        N_nearest_neighbours = min_dist_neighbours[:N]
+        features["average_dist_N_nearest_neighbours_pixel"] = sum(
+            N_nearest_neighbours
+        ) / len(N_nearest_neighbours)
+        features["average_dist_N_nearest_neighbours_rel"] = (
+            features["average_dist_N_nearest_neighbours_pixel"] / invoice_diag
         )
 
-        ### Invoice number alignment related features ###
-        # TODO: Implement features that determine if token is aligned with invoice number related tokens
+        # relative size compared to other tokens (percentile)
+        perc_w = 0
+        perc_h = 0
+        for t in invoicePage.grouped_tokens:
+            if t.coordinates["width"] < token.coordinates["width"]:
+                perc_w += 1
+            if t.coordinates["height"] < token.coordinates["height"]:
+                perc_h += 1
+        features["percentile_width"] = perc_w / len(invoicePage.grouped_tokens)
+        features["percentile_height"] = perc_h / len(invoicePage.grouped_tokens)
 
-        ### Tokens nearby ###
-        # TODO: Use nearby tokens as feature
+        # boolean if token contains fields
+        features["contains_date"] = True if token.date_values else False
+        features["contains_currency"] = True if token.currency else False
+        features["contains_address"] = True if token.address else False
+        features["contains_num_label"] = True if token.num_lable else False
+        features["contains_total_label"] = True if token.total_label else False
+        features["contains_date_label"] = True if token.date_label else False
+        features["contains_digit"] = True if token.contains_digit else False
 
-        return features
+        # boolean if aligned with selected tokens
+        moe = 10  # arbitary 10 pixle margin of error
+        features["vert_align_to_cell_w_date"] = False
+        features["vert_align_to_cell_w_currency"] = False
+        features["vert_align_to_cell_w_address"] = False
+        features["vert_align_to_cell_w_datelabel"] = False
+        features["vert_align_to_cell_w_numlabel"] = False
+        features["vert_align_to_cell_w_totallabel"] = False
+        features["vert_align_to_cell_w_digit"] = False
+
+        for t in invoicePage.grouped_tokens:
```

Comment:
Seems like this would also include the token that is currently being examined, so if it contains a date label it would return True even though it is only aligned to itself. Perhaps filter out the current token from grouped tokens before doing the for loop?

Reply:
solved